### PR TITLE
[Inductor][FlexAttention] Cast load_checked_2d offsets to INDEX_DTYPE (avoid AMD int32 element-offset overflow)

### PR DESCRIPTION
**Pull request by haoyuz**
Summary:
Compiled FlexAttention faults on AMD MI350X when a paged KV cache's K/V load element offset exceeds INT32_MAX. The K-load inside `load_checked_2d` computes `offs_m[:, None] * stride_m + offs_n[None, :] * stride_n` with int32 operands. For paged KV caches with > 131072 16-token pages and `num_kv_heads=8, head_dim=128`, the per-token element stride is 1024 and the highest page's K-load reaches an offset of `(131072 * 16 + 15) * 1024 = 2,147,498,015` elements — past INT32_MAX (`2,147,483,647`). The i32 multiply wraps and the load dereferences invalid memory: on AMD this either traps (the original `flex_impl_nan_repro_op_min` symptom) or silently returns garbage from whatever page the wrapped 32-bit byte offset lands on. NVIDIA's PTX uses plain 64-bit `ld.global.*` and is unaffected at the same shape.

This diff plumbs the existing Inductor `INDEX_DTYPE` constexpr (already set by `TritonTemplate.maybe_append_choice` via `can_use_32bit_indexing`) through `load_checked_2d` and every flex template callsite, then casts the offset operands to it before multiplying by the strides. For our failing shape Inductor's element-count heuristic picks `INDEX_DTYPE = tl.int64`, and the cast promotes the i32 offsets to i64 so the multiply does not wrap. Mirrors the existing `.to(tl.int64)` casts in `flex_backwards.py.jinja`. No behaviour change on NVIDIA.

Credit: Taylor Robie (D108082283) for the INDEX_DTYPE plumbing idea. Earlier versions of this diff also added a byte-size-aware override on top of Inductor's element-count heuristic, on the suspicion that the AMD buffer-resource NUM_RECORDS cap is byte-indexed and would trip for `elements < INT32_MAX, bytes > 2 GiB`. Empirical testing on MI350X (`num_blocks=131071`, K elements `2.147 G < INT32_MAX`, K bytes `4.29 GiB > 2 GiB`) shows that case runs cleanly across 5 consecutive runs — Triton's AMD backend emits a typed `buffer_load_short` (for bf16) whose addressing is element-based, so the byte size does not matter. The byte-size override has been removed; Inductor's element-count heuristic is sufficient.

`caffe2/test/inductor/BUCK`: makes the `test_aot_inductor_model_runner_pybind` dep on `:test_aot_inductor_utils` conditional on CUDA (the pybind includes `c10/cuda/CUDAStream.h` and is not wired through the hipify gen, so it doesn't compile under the AMD toolchain; its Python-side import is function-scoped so non-AOTI tests don't trigger it). Without this BUCK fix the regression test below cannot build on AMD.

Adds regression test `caffe2/test/inductor/test_flex_attention.py::TestFlexAttention::test_paged_kv_load_int32_element_overflow`. Shape (`num_blocks=131073`, `num_kv_heads=8`, `head_dim=128`, `block_size=16`, bf16, `dynamic=True`) is the original production repro: max K-load element offset = `2,147,499,008 > INT32_MAX`, Inductor picks `INDEX_DTYPE = tl.int64`, the `.to(INDEX_DTYPE)` cast in `load_checked_2d` is what protects the load. The test compares the compiled output against a `scaled_dot_product_attention` reference computed over only the in-mask block, which catches both failure modes (hard trap and silent wrong-page garbage even when finite).

Test Plan:
PyTorch unit test on AMD MI350X (`devgpu153.prn5`, ROCm 7.0, gfx950):

```
$ HIP_VISIBLE_DEVICES=7 buck2 run @//mode/opt -m fbcode//vllm:omni @//mode/opt-amd-gpu \
    -c fbcode.rocm_arch=mi350 -m rocm70 fbcode//caffe2/test/inductor:flex_attention -- \
    caffe2.test.inductor.test_flex_attention.TestFlexAttentionCUDA.test_paged_kv_load_int32_element_overflow_cuda
```

With the cast applied (this diff):
```
test_paged_kv_load_int32_element_overflow_cuda ... ok
Ran 1 test in 2.357s
OK
```

With the `.to(INDEX_DTYPE)` cast in `load_checked_2d` temporarily removed (simulating "pre-fix" state):
```
test_paged_kv_load_int32_element_overflow_cuda ... Memory access fault by GPU node-9 (Agent handle: 0x7f94d33d3800) on address 0x7f4e11606000. Reason: Unknown.
exit=1
```

Falsifiability check for the dropped byte-size override (5 consecutive runs on MI350X, `num_blocks=131071` — K elements 2.147 G < INT32_MAX, K bytes 4.29 GiB > 2 GiB, no override): all 5 passed. AMD's typed buffer_load handles itemsize internally; element count is the right metric for the cap.

Production users have independently confirmed the original `flex_impl_nan_repro_op_min` repro from the Workplace report (https://fb.workplace.com/groups/729207668438473/permalink/1881778503181378/) is fixed.

H100 build and run with the same source produces no fault either way (no behaviour change on NVIDIA).

Reviewed By: jinzhenfan, hyuen

Differential Revision: D107927427

@diff-train-skip-merge

### CODE REVIEW

File: torch/_inductor/kernel/flex/templates/utilities.py.jinja [hunk 1]
```diff
@@ -43,10 +43,19 @@ def load_checked_2d(
...
     M_LEN: tl.constexpr,
     N_LEN: tl.constexpr,
+    INDEX_DTYPE: tl.constexpr = tl.int32,
 ):
-    # Calculate final pointer if strides are provided
+    # Calculate final pointer if strides are provided. The caller controls
+    # the index dtype via INDEX_DTYPE; Inductor picks int64 whenever the
+    # source tensor's byte size could exceed the AMD buffer-resource
+    # NUM_RECORDS cap (~2 GiB), which would otherwise truncate the load
+    # offset to a 32-bit value at SASS emission and silently fault.
```
Comment (robieta) [inline torch/_inductor/kernel/flex/templates/utilities.py.jinja]:
```suggestion
    # offset to a 32-bit value and silently fault.
```
Nit: SASS is Nvida only.

File: torch/_inductor/select_algorithm.py [hunk 2]
```diff
@@ -204,9 +204,9 @@ def code(self) -> str:
         remaining_active_hooks = [
             key for key, fn in self.replacement_hooks.items() if fn is not None
         ]
-        assert len(remaining_active_hooks) == 0, (
-            f"The following hooks have not yet been finalized:\n {remaining_active_hooks=}"
-        )
+        assert (
```
Comment (robieta) [inline torch/_inductor/select_algorithm.py]: Looks like some issue around export since I don't see these formatting changes in the diff?

File: torch/_inductor/kernel/flex/flex_attention.py [hunk 3]
```diff
@@ -107,6 +107,36 @@ def get_float32_precision():
...
+# exceeds 2 GiB while their element count still fits in 32 bits (e.g. a bf16
+# paged KV cache with ~1.6 G elements = 3.2 GiB). For such tensors the load
+# offset can overrun the buffer-resource cap and silently fault.
+#
+# This helper returns "tl.int64" whenever any of the supplied buffers'
+# byte sizes cannot be statically proven to fit in 2 GiB, regardless of
+# element count. The caller threads the result into `cur_kernel_options`
+# as `INDEX_DTYPE`, overriding the element-count-based default set by
+# `TritonTemplate.maybe_append_choice`.
+def _flex_index_dtype_for_buffers(*buffers) -> str:
```
Comment (robieta) [inline torch/_inductor/kernel/flex/flex_attention.py]: Why is the standard Inductor `TritonScheduling.can_use_32bit_indexing` not sufficient? If it's just a question of elements vs. bytes doesn't that imply that we need to fix Inductor since everything would be broken on AMD?

File: torch/_inductor/select_algorithm.py (hunk 2 above)
Comment (haoyuz) [reply torch/_inductor/select_algorithm.py]: There were some issues in auto formatting. Reverted, thanks!

File: torch/_inductor/kernel/flex/flex_attention.py (hunk 3 above)
Comment (haoyuz) [reply torch/_inductor/kernel/flex/flex_attention.py]: OK seems that my Claude firmly believes that can_use_32bit_indexing only checks for # of elements, but here we need the # of bytes to fit in int32 max. Checking the inductor IR's code, I do see that it uses `compute_required_storage_length` which returns the element count. Please let me know if this makes sense to you :)

**Review by hyuen — APPROVED** [review]

**Review by scxiao — APPROVED** [review]

**Comment by scxiao** [comment]
Hi @haoyuz, @robieta, I did some investigation on the kernel, I think the problem here is not related to the `buffer_load`. The `buffer_load` in the added test is only for `q` input, `k/v` loads actually use `global load`. You can see the `ttgir` as:

```
      %k_157 = tt.load %ptr_153, %k_156, %cst_2 : tensor<16x128x!tt.ptr<bf16>, #blocked> loc(#loc252)
      %v_190 = tt.load %ptr_189, %k_156, %cst_2 : tensor<16x128x!tt.ptr<bf16>, #blocked> loc(#loc253)
```
and the ISA calls `buffer_load_dwordx4` only for `q` input, no matter whether we cast `offs_m` and `offs_n` to `tl.int64`.

I think the root cause is that `offs_m[:, None] * stride_m` causes overflow for int32 due to the large value of `stride_m`, so  we need to cast it to `int64` in the kernel side, and the changes in this PR are the right fix. I played with the changes, if I only add cast to int64 for `offs_n`, i.e., the changes in `load_checked_2d` look like:
```
        ptr = (
            ptr
            + offs_m[:, None] * stride_m
            + offs_n[None, :].to(INDEX_DTYPE) * stride_n
        )
``` 
The test still failed. If you do the cast to `offs_m`, i.e., the changes are:
```
        ptr = (
            ptr
            + offs_m[:, None].to(INDEX_DTYPE)  * stride_m
            + offs_n[None, :]* stride_n
        )
```
Then the test passes. 

As for whether lowering a load to `buffer_load`, we have a check at: https://github.com/triton-lang/triton/blob/1859e68dbbfa746c31adc44ddd399975bc03c91b/third_party/amd/backend/compiler.py#L224. This check ensures if the input tensor size is larger than 2GB, we will not lower it to `buffer_load`. 

Finally, I agree that set the `num_records` to `MAX_INT32` in buffer load has some issues, I think it should be the bytes from the load_ptr to the end of the buffer.

**Comment by drisspg** [comment]
https://github.com/pytorch/pytorch/pull/185264
supersedes this right?
